File: regression/rt_5m_hartmann.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <gkyl_alloc.h>
#include <gkyl_moment.h>
#include <gkyl_util.h>
#include <gkyl_wv_euler.h>

#include <gkyl_null_comm.h>

#ifdef GKYL_HAVE_MPI
#include <mpi.h>
#include <gkyl_mpi_comm.h>
#endif

#include <rt_arg_parse.h>

struct hartmann_ctx
{
  // Mathematical constants (dimensionless).
  double pi;

  // Physical constants (using normalized code units).
  double gas_gamma; // Adiabatic index.
  double epsilon0; // Permittivity of free space.
  double mu0; // Permeability of free space.
  double mass_ion; // Ion mass.
  double charge_ion; // Ion charge.
  double mass_elc; // Electron mass.
  double charge_elc; // Electron charge.

  double n0; // Reference number density.
  double coll_fac; // Collision factor.

  double beta; // Plasma beta.
  double grav; // Gravitational acceleration.

  // Derived physical quantities (using normalized code units).
  double light_speed; // Speed of light.
  double vAe; // Electron Alfven velocity.
  double B0; // Reference magnetic field strength.
  double omega_pi; // Ion plasma frequency.
  double di; // Ion skin depth.

  double T_elc; // Electron temperature.
  double T_ion; // Ion temperature.

  double rho_elc; // Electron mass density.
  double rho_ion; // Ion mass density.

  double E_elc; // Electron total energy density.
  double E_ion; // Ion total energy density.

  double tau; // Collision time.
  double lambda; // Collision wavelength.

  // Simulation parameters.
  int Nx; // Cell count (x-direction).
  double Lx; // Domain size (x-direction).
  double cfl_frac; // CFL coefficient.

  double t_end; // Final simulation time.
  int num_frames; // Number of output frames.
  double dt_failure_tol; // Minimum allowable fraction of initial time-step.
  int num_failures_max; // Maximum allowable number of consecutive small time-steps.
};

struct hartmann_ctx
create_ctx(void)
{
  // Mathematical constants (dimensionless).
  double pi = M_PI;

  // Physical constants (using normalized code units).
  double gas_gamma = 5.0 / 3.0; // Adiabatic index.
  double epsilon0 = 1.0; // Permittivity of free space.
  double mu0 = 1.0; // Permeability of free space.
  double mass_ion = 1.0; // Ion mass.
  double charge_ion = 1.0; // Ion charge.
  double mass_elc = 1.0 / sqrt(1836.153); // Electron mass.
  double charge_elc = -1.0; // Electron charge.

  double n0 = 1.0; // Reference number density.
  double coll_fac = 1e5; // Collision factor.

  double beta = 0.1; // Plasma beta.
  double grav = 0.01; // Gravitational acceleration.

  // Derived physical quantities (using normalized code units).
  double light_speed = 1.0 / sqrt(epsilon0 * mu0); // Speed of light.
  double vAe = light_speed; // Electron Alfven velocity.
  double B0 = vAe * sqrt(mu0 * n0 * mass_elc); // Reference magnetic field strength.
  double omega_pi = sqrt(charge_ion * charge_ion * n0 / (epsilon0 * mass_ion)); // Ion plasma frequency.
  double di = light_speed / omega_pi; // Ion skin depth.
  
  double T_elc = beta * (B0 * B0) / (2.0 * n0 * mu0); // Electron temperature.
  double T_ion = beta * (B0 * B0) / (2.0 * n0 * mu0); // Ion temperature.

  double rho_elc = n0 * mass_elc; // Electron mass density.
  double rho_ion = n0 * mass_ion; // Ion mass density.

  double E_elc = n0 * T_elc / (gas_gamma - 1.0); // Electron total energy density.
  double E_ion = n0 * T_ion / (gas_gamma - 1.0); // Ion total energy density.

  double tau = coll_fac * 6.0 * sqrt(2.0 * pi * mass_elc * T_elc * pi * T_elc * pi * T_elc) * epsilon0 * epsilon0 /
    (charge_ion * charge_ion * charge_ion * charge_ion * n0); // Collision time.
  double lambda = 1.0 / mu0 * (mass_elc / (charge_ion * charge_ion * n0 * tau)); // Collision wavelength.

  // Simulation parameters.
  int Nx = 256; // Cell count (x-direction).
  double Lx = 256.0 * di; // Domain size (x-direction).
  double cfl_frac = 0.001; // CFL coefficient.

  double t_end = 1.0 * tau; // Final simulation time.
  int num_frames = 1; // Number of output frames.
  double dt_failure_tol = 1.0e-4; // Minimum allowable fraction of initial time-step.
  int num_failures_max = 20; // Maximum allowable number of consecutive small time-steps.

  struct hartmann_ctx ctx = {
    .pi = pi,
    .gas_gamma = gas_gamma,
    .epsilon0 = epsilon0,
    .mu0 = mu0,
    .mass_ion = mass_ion,
    .charge_ion = charge_ion,
    .mass_elc = mass_elc,
    .charge_elc = charge_elc,
    .n0 = n0,
    .coll_fac = coll_fac,
    .beta = beta,
    .grav = grav,
    .light_speed = light_speed,
    .vAe = vAe,
    .B0 = B0,
    .omega_pi = omega_pi,
    .di = di,
    .T_elc = T_elc,
    .T_ion = T_ion,
    .rho_elc = rho_elc,
    .rho_ion = rho_ion,
    .E_elc = E_elc,
    .E_ion = E_ion,
    .tau = tau,
    .lambda = lambda,
    .Nx = Nx,
    .Lx = Lx,
    .cfl_frac = cfl_frac,
    .t_end = t_end,
    .num_frames = num_frames,
    .dt_failure_tol = dt_failure_tol,
    .num_failures_max = num_failures_max,
  };

  return ctx;
}

void
evalElcInit(double t, const double* GKYL_RESTRICT xn, double* GKYL_RESTRICT fout, void* ctx)
{
  struct hartmann_ctx *app = ctx;

  double rho_elc = app->rho_elc;
  double E_elc = app->E_elc;

  // Set electron mass density.
  fout[0] = rho_elc;
  // Set electron momentum density.
  fout[1] = 0.0; fout[2] = 0.0; fout[3] = 0.0;
  // Set electron total energy density.
  fout[4] = E_elc;
}

void
evalIonInit(double t, const double* GKYL_RESTRICT xn, double* GKYL_RESTRICT fout, void* ctx)
{
  struct hartmann_ctx *app = ctx;

  double rho_ion = app->rho_ion;
  double E_ion = app->E_ion;

  // Set ion mass density.
  fout[0] = rho_ion;
  // Set ion momentum density.
  fout[1] = 0.0; fout[2] = 0.0; fout[3] = 0.0;
  // Set ion total energy density.
  fout[4] = E_ion;
}

void
evalFieldInit(double t, const double* GKYL_RESTRICT xn, double* GKYL_RESTRICT fout, void* ctx)
{
  struct hartmann_ctx *app = ctx;

  double B0 = app->B0;

  // Set electric field.
  fout[0] = 0.0, fout[1] = 0.0; fout[2] = 0.0;
  // Set magnetic field.
  fout[3] = B0, fout[4] = 0.0; fout[5] = 0.0;
  // Set correction potentials.
  fout[6] = 0.0; fout[7] = 0.0;
}

void
evalAppAccel(double t, const double * GKYL_RESTRICT xn, double* GKYL_RESTRICT fout, void *ctx)
{
  struct hartmann_ctx *app = ctx;

  double grav = app->grav;

  // Set applied acceleration.
  fout[0] = 0.0; fout[1] = grav; fout[2] = 0.0;
}

void
write_data(struct gkyl_tm_trigger* iot, gkyl_moment_app* app, double t_curr, bool force_write)
{
  if (gkyl_tm_trigger_check_and_bump(iot, t_curr)) {
    int frame = iot->curr - 1;
    if (force_write) {
      frame = iot->curr;
    }

    gkyl_moment_app_write(app, t_curr, frame);
  }
}

int
main(int argc, char **argv)
{
  struct gkyl_app_args app_args = parse_app_args(argc, argv);

#ifdef GKYL_HAVE_MPI
  if (app_args.use_mpi) {
    MPI_Init(&argc, &argv);
  }
#endif

  if (app_args.trace_mem) {
    gkyl_cu_dev_mem_debug_set(true);
    gkyl_mem_debug_set(true);
  }

  struct hartmann_ctx ctx = create_ctx(); // Context for initialization functions.

  int NX = APP_ARGS_CHOOSE(app_args.xcells[0], ctx.Nx);

  // Electron/ion equations.
  struct gkyl_wv_eqn *elc_euler = gkyl_wv_euler_new(ctx.gas_gamma, app_args.use_gpu);
  struct gkyl_wv_eqn *ion_euler = gkyl_wv_euler_new(ctx.gas_gamma, app_args.use_gpu);

  struct gkyl_moment_species elc = {
    .name = "elc",
    .charge = ctx.charge_elc, .mass = ctx.mass_elc,
    .equation = elc_euler,
    .evolve = true,
    .init = evalElcInit,
    .app_accel_func = evalAppAccel,
    .ctx = &ctx,

    .type_brag = GKYL_BRAG_MAG_FULL,

    .bcx = { GKYL_SPECIES_NO_SLIP, GKYL_SPECIES_NO_SLIP },
  };

  struct gkyl_moment_species ion = {
    .name = "ion",
    .charge = ctx.charge_ion, .mass = ctx.mass_ion,
    .equation = ion_euler,
    .evolve = true,
    .init = evalIonInit,
    .app_accel_func = evalAppAccel,
    .ctx = &ctx,

    .type_brag = GKYL_BRAG_MAG_FULL,

    .bcx = { GKYL_SPECIES_NO_SLIP, GKYL_SPECIES_NO_SLIP },
  };

  // Field.
  struct gkyl_moment_field field = {
    .epsilon0 = ctx.epsilon0, .mu0 = ctx.mu0,
    
    .evolve = false,
    .init = evalFieldInit,
    .ctx = &ctx,

    .bcx = { GKYL_FIELD_PEC_WALL, GKYL_FIELD_PEC_WALL },
  };

  int nrank = 1; // Number of processes in simulation.
#ifdef GKYL_HAVE_MPI
  if (app_args.use_mpi) {
    MPI_Comm_size(MPI_COMM_WORLD, &nrank);
  }
#endif

  // Create global range.
  int cells[] = { NX };
  int dim = sizeof(cells) / sizeof(cells[0]);
  struct gkyl_range global_r;
  gkyl_create_global_range(dim, cells, &global_r);

  // Create decomposition.
  int cuts[dim];
#ifdef GKYL_HAVE_MPI
  for (int d = 0; d < dim; d++) {
    if (app_args.use_mpi) {
      cuts[d] = app_args.cuts[d];
    }
    else {
      cuts[d] = 1;
    }
  }
#else
  for (int d = 0; d < dim; d++) {
    cuts[d] = 1;
  }
#endif

  struct gkyl_rect_decomp *decomp = gkyl_rect_decomp_new_from_cuts(dim, cuts, &global_r);

  // Construct communicator for use in app.
  struct gkyl_comm *comm;
#ifdef GKYL_HAVE_MPI
  if (app_args.use_mpi) {
    comm = gkyl_mpi_comm_new( &(struct gkyl_mpi_comm_inp) {
        .mpi_comm = MPI_COMM_WORLD,
        .decomp = decomp
      }
    );
  }
  else {
    comm = gkyl_null_comm_inew( &(struct gkyl_null_comm_inp) {
        .decomp = decomp,
        .use_gpu = app_args.use_gpu
      }
    );
  }
#else
  comm = gkyl_null_comm_inew( &(struct gkyl_null_comm_inp) {
      .decomp = decomp,
      .use_gpu = app_args.use_gpu
    }
  );
#endif

  int my_rank;
  gkyl_comm_get_rank(comm, &my_rank);
  int comm_size;
  gkyl_comm_get_size(comm, &comm_size);

  int ncuts = 1;
  for (int d = 0; d < dim; d++) {
    ncuts *= cuts[d];
  }

  if (ncuts != comm_size) {
    if (my_rank == 0) {
      fprintf(stderr, "*** Number of ranks, %d, does not match total cuts, %d!\n", comm_size, ncuts);
    }
    goto mpifinalize;
  }

  // Moment app.
  struct gkyl_moment app_inp = {
    .name = "5m_hartmann",

    .ndim = 1,
    .lower = { 0.0 },
    .upper = { ctx.Lx },
    .cells = { NX },

    .num_periodic_dir = 0,
    .periodic_dirs = { },
    .cfl_frac = ctx.cfl_frac,

    .has_braginskii = true,
    .coll_fac = ctx.coll_fac,

    .num_species = 2,
    .species = { elc, ion },

    .field = field,

    .parallelism = {
      .use_gpu = app_args.use_gpu,
      .cuts = { app_args.cuts[0] },
      .comm = comm,
    },
  };

  // Create app object.
  gkyl_moment_app *app = gkyl_moment_app_new(&app_inp);

  // Initial and final simulation times.
  double t_curr = 0.0, t_end = ctx.t_end;

  // Create trigger for IO.
  int num_frames = ctx.num_frames;
  struct gkyl_tm_trigger io_trig = { .dt = t_end / num_frames };

  // Initialize simulation.
  gkyl_moment_app_apply_ic(app, t_curr);
  write_data(&io_trig, app, t_curr, false);

  // Compute estimate of maximum stable time-step.
  double dt = gkyl_moment_app_max_dt(app);

  // Initialize small time-step check.
  double dt_init = -1.0, dt_failure_tol = ctx.dt_failure_tol;
  int num_failures = 0, num_failures_max = ctx.num_failures_max;

  long step = 1;
  while ((t_curr < t_end) && (step <= app_args.num_steps)) {
    gkyl_moment_app_cout(app, stdout, "Taking time-step %ld at t = %g ...", step, t_curr);
    struct gkyl_update_status status = gkyl_moment_update(app, dt);
    gkyl_moment_app_cout(app, stdout, " dt = %g\n", status.dt_actual);
    
    if (!status.success) {
      gkyl_moment_app_cout(app, stdout, "** Update method failed! Aborting simulation ....\n");
      break;
    }

    t_curr += status.dt_actual;
    dt = status.dt_suggested;

    write_data(&io_trig, app, t_curr, false);

    if (dt_init < 0.0) {
      dt_init = status.dt_actual;
    }
    else if (status.dt_actual < dt_failure_tol * dt_init) {
      num_failures += 1;

      gkyl_moment_app_cout(app, stdout, "WARNING: Time-step dt = %g", status.dt_actual);
      gkyl_moment_app_cout(app, stdout, " is below %g*dt_init ...", dt_failure_tol);
      gkyl_moment_app_cout(app, stdout, " num_failures = %d\n", num_failures);
      if (num_failures >= num_failures_max) {
        gkyl_moment_app_cout(app, stdout, "ERROR: Time-step was below %g*dt_init ", dt_failure_tol);
        gkyl_moment_app_cout(app, stdout, "%d consecutive times. Aborting simulation ....\n", num_failures_max);
        break;
      }
    }
    else {
      num_failures = 0;
    }

    step += 1;
  }

  write_data(&io_trig, app, t_curr, false);
  gkyl_moment_app_stat_write(app);

  struct gkyl_moment_stat stat = gkyl_moment_app_stat(app);

  gkyl_moment_app_cout(app, stdout, "\n");
  gkyl_moment_app_cout(app, stdout, "Number of update calls %ld\n", stat.nup);
  gkyl_moment_app_cout(app, stdout, "Number of failed time-steps %ld\n", stat.nfail);
  gkyl_moment_app_cout(app, stdout, "Species updates took %g secs\n", stat.species_tm);
  gkyl_moment_app_cout(app, stdout, "Field updates took %g secs\n", stat.field_tm);
  gkyl_moment_app_cout(app, stdout, "Source updates took %g secs\n", stat.sources_tm);
  gkyl_moment_app_cout(app, stdout, "Total updates took %g secs\n", stat.total_tm);

  // Free resources after simulation completion.
  gkyl_wv_eqn_release(elc_euler);
  gkyl_wv_eqn_release(ion_euler);
  gkyl_rect_decomp_release(decomp);
  gkyl_comm_release(comm);
  gkyl_moment_app_release(app);  
  
mpifinalize:
#ifdef GKYL_HAVE_MPI
  if (app_args.use_mpi) {
    MPI_Finalize();
  }
#endif
  
  return 0;
}

File: regression/lua/rt_5m_hartmann.lua
local Moments = G0.Moments
local Euler = G0.Moments.Eq.Euler

-- Mathematical constants (dimensionless).
pi = math.pi

-- Physical constants (using normalized code units).
gas_gamma = 5.0 / 3.0 -- Adiabatic index.
epsilon0 = 1.0 -- Permittivity of free space.
mu0 = 1.0 -- Permeability of free space.
mass_ion = 1.0 -- Ion mass.
charge_ion = 1.0 -- Ion charge.
mass_elc = 1.0 / math.sqrt(1836.153) -- Electron mass.
charge_elc = -1.0 -- Electron charge.

n0 = 1.0 -- Reference number density.
coll_fac = 1e5 -- Collision factor.

beta = 0.1 -- Plasma beta.
grav = 0.01 -- Gravitational acceleration.

-- Derived physical quantities (using normalized code units).
light_speed = 1.0 / math.sqrt(epsilon0 * mu0) -- Speed of light.
vAe = light_speed -- Electron Alfven velocity.
B0 = vAe * math.sqrt(mu0 * n0 * mass_elc) -- Reference magnetic field strength.
omega_pi = math.sqrt(charge_ion * charge_ion * n0 / (epsilon0 * mass_ion)) -- Ion plasma frequency.
di = light_speed / omega_pi -- Ion skin depth.

T_elc = beta * (B0 * B0) / (2.0 * n0 * mu0) -- Electron temperature.
T_ion = beta * (B0 * B0) / (2.0 * n0 * mu0) -- Ion temperature.

rho_elc = n0 * mass_elc -- Electron mass density.
rho_ion = n0 * mass_ion -- Ion mass density.

E_elc = n0 * T_elc / (gas_gamma - 1.0) -- Electron total energy density.
E_ion = n0 * T_ion / (gas_gamma - 1.0) -- Ion total energy density.

tau = coll_fac * 6.0 * math.sqrt(2.0 * pi * mass_elc * T_elc * pi * T_elc * pi * T_elc) * epsilon0 * epsilon0 /
  (charge_ion * charge_ion * charge_ion * charge_ion * n0) -- Collision time.
lambda = 1.0 / mu0 * (mass_elc / (charge_ion * charge_ion * n0 * tau)) -- Collision wavelength.

-- Simulation parameters.
Nx = 256 -- Cell count (x-direction).
Lx = 256.0 * di -- Domain size (x-direction).
cfl_frac = 0.001 -- CFL coefficient.

t_end = 1.0 * tau -- Final simulation time.
num_frames = 1 -- Number of output frames.
field_energy_calcs = GKYL_MAX_INT -- Number of times to calculate field energy.
integrated_mom_calcs = GKYL_MAX_INT -- Number of times to calculate integrated moments.
dt_failure_tol = 1.0e-4 -- Minimum allowable fraction of initial time-step.
num_failures_max = 20 -- Maximum allowable number of consecutive small time-steps.

momentApp = Moments.App.new {
  
  tEnd = t_end,
  nFrame = num_frames,
  fieldEnergyCalcs = field_energy_calcs,
  integratedMomentCalcs = integrated_mom_calcs,
  dtFailureTol = dt_failure_tol,
  numFailuresMax = num_failures_max,
  lower = { 0.0 },
  upper = { Lx },
  cells = { Nx },
  cflFrac = cfl_frac,

  hasBraginskii = true,
  collisionFactor = coll_fac,
    
  -- Boundary conditions for configuration space.
  periodicDirs = { }, -- Periodic directions (none).
  
  -- Electrons.
  elc = Moments.Species.new {
    charge = charge_elc, mass = mass_elc,
    equation = Euler.new { gasGamma = gas_gamma },
    braginskiiType = G0.Braginskii.MagFull,
    
    -- Initial conditions function.
    init = function (t, xn)
      local rhoe = rho_elc -- Electron mass density.
      local mome_x = 0.0 -- Electron momentum density (x-direction).
      local mome_y = 0.0 -- Electron momentum density (y-direction).
      local mome_z = 0.0 -- Electron momentum density (z-direction).
      local Ee_tot = E_elc -- Electron total energy density.
      
      return rhoe, mome_x, mome_y, mome_z, Ee_tot
    end,

    -- Applied acceleration function.
    appliedAcceleration = function (t, xn)
      local accele_x = 0.0 -- Electron applied acceleration (x-direction).
      local accele_y = grav -- Electron applied acceleration (y-direction).
      local accele_z = 0.0 -- Electron applied acceleration (z-direction).

      return accele_x, accele_y, accele_z
    end,
  
    evolve = true, -- Evolve species?
    bcx = { G0.SpeciesBc.bcNoSlip, G0.SpeciesBc.bcNoSlip } -- No-slip boundary conditions (x-direction).
  },

  -- Ions.
  ion = Moments.Species.new {
    charge = charge_ion, mass = mass_ion,
    equation = Euler.new { gasGamma = gas_gamma },
    braginskiiType = G0.Braginskii.MagFull,
    
    -- Initial conditions function.
    init = function (t, xn)
      local rhoi = rho_ion -- Ion mass density.
      local momi_x = 0.0 -- Ion momentum density (x-direction).
      local momi_y = 0.0 -- Ion momentum density (y-direction).
      local momi_z = 0.0 -- Ion momentum density (z-direction).
      local Ei_tot = E_ion -- Ion total energy density.
      
      return rhoi, momi_x, momi_y, momi_z, Ei_tot
    end,

    -- Applied acceleration function.
    appliedAcceleration = function (t, xn)
      local acceli_x = 0.0 -- Ion applied acceleration (x-direction).
      local acceli_y = grav -- Ion applied acceleration (y-direction).
      local acceli_z = 0.0 -- Ion applied acceleration (z-direction).
  
      return acceli_x, acceli_y, acceli_z
    end,
  
    evolve = true, -- Evolve species?
    bcx = { G0.SpeciesBc.bcNoSlip, G0.SpeciesBc.bcNoSlip } -- No-slip boundary conditions (x-direction).
  },
    
  -- Field.
  field = Moments.Field.new {
    epsilon0 = epsilon0, mu0 = mu0,

    -- Initial conditions function.
    init = function (t, xn)
      local Ex = 0.0 -- Total electric field (x-direction).
      local Ey = 0.0 -- Total electric field (y-direction).
      local Ez = 0.0 -- Total electric field (z-direction).

      local Bx = B0 -- Total magnetic field (x-direction).
      local By = 0.0 -- Total magnetic field (y-direction).
      local Bz = 0.0 -- Total magnetic field (z-direction).
      
      return Ex, Ey, Ez, Bx, By, Bz, 0.0, 0.0
    end,

    evolve = false, -- Evolve field?
    bcx = { G0.FieldBc.bcPECWall, G0.FieldBc.bcPECWall } -- PEC wall boundary conditions (x-direction).
  }
}

-- Run application.
momentApp:run()
